### ISSUE TITLE
LogiX Node: On DeFocus World

### ISSUE THREAD

**Issue by seif1** [Neos-Metaverse/NeosPublic#2391]
## Is your feature request related to a problem? Please describe.
I would like to write a cloud variable whenever a user focuses, or joins a different world so that it saves the value it was in the world they just left so that value can be accessed in the newly joined world.

There is currently no way that I could find to do this without having to constantly update the variable.

## Relevant issues
None that I found

## Describe the solution you'd like
A LogiX node which takes a user input and gives a impulse as that user when they defocus the world

## Describe alternatives you've considered
Possibly some component that can be used in this manner with cloud variables to write to them whenever the variable owner leaves or Defocuses a world, instead of a LogiX node.

## Additional context
Just a idea, could be many different ways to get a similar or more robust solution.

**Comment by H3BO3** [comment]
This would probably have a user output, if anything, like the user join/leave nodes. It would cause coding inefficiencies to have to provide a user.

Another node could also be added for a world Focus event as well.

**Comment by seif1** [comment]
That is probably a better method of implementation, keeping things more in line with existing nodes that have a similar function

**Comment by KyuubiYoru** [comment]
A node for Focus and Defocus with Impulse and User Output like the "User Spawned" node would be great.

**Comment by Frooxius** [comment]
What exactly are you trying to do with this?

We can provide the nodes for the events themselves, but that's a completely separate system from cloud variables, so I'm not too sure about connecting them in some way. It sounds like you might be trying to do something with those instead?

**Comment by seif1** [comment]
My goals:
I am trying to make a avatar system which can save data across worlds, and neos sessions (when the game is closed), without the user having to actually re-save the avatar, and spawn the new copy every time they change worlds, with the end goal of whatever world the user is in always having the most up to date data available without requiring any action from them. 
Ideally, it would also save data separately if the user is using a different avatar entirely, but I understand that is more demanding than just saving 1 global copy which is used across all avatars that have this system.

My thoughts on how this node would help what I am trying to accomplish:
When the user focuses into a different world, the most recent data from their avatar would come with them because it is written to a cloud variable when they start to focus a different world, and read in that world they focus into. This would preferably be changed to some sort of entirely local data system if/when that is implemented (seems to be planned for the Logix Rework) because this data is only saved to allow the user to transfer it between worlds/neos sessions without having to save the avatar again. 

I may have been too focused on the solution and not enough on the problem, if there is some other method that may be implemented in this case which would better facilitate what I am trying to accomplish, that would be wonderful as well.

**Comment by shadowpanther** [comment]
What stops you from just writing every ongoing change into the cloud variable immediately? The avatar in the next focused world would just receive the update.

**Comment by seif1** [comment]
> What stops you from just writing every ongoing change into the cloud variable immediately? The avatar in the next focused world would just receive the update.

That seems extremely excessive considering I am driving it within the world, and it is constantly changing (I have one value that is the amount the main value changes per second, and I am trying to sync the main value), worst case I would a timer every so often to trigger a update, but would prefer to just sync the data when it is actually relevant (when the user is exiting the game, changing avatars, respawning, or changing worlds)

**Comment by shiftyscales** [comment]
Yeah, there's not really any reason that you can't continually write the changes. The newest data will always be pulled when it can be, @seif1. It's not really excessive if the data needs to be continually changing.

**Comment by seif1** [comment]
> Yeah, there's not really any reason that you can't continually write the changes. The newest data will always be pulled when it can be, @seif1. It's not really excessive if the data needs to be continually changing.

I would eventually like to shift the system so it does not use cloud variables (if there is ever a local version of "cloud" variables that are saved only by the user, and aren't synced to sessions they are not in, which appears to be planned in the future from my understanding), so depending on that implementation, it would mean having to restructure the entire sync system, rather than just building it to be future proofed in the first place, so while that may be acceptable for cloud variables, because they don't sync often in worlds the user is not in, it may be less ideal in any later implementation that the system would be switched to.

**Comment by Frooxius** [comment]
I think you might be trying to solve a problem that doesn't really exist. Neos already synchronizes cloud variables across sessions in realtime for the user, so simply writing to it, or using the cloud variable components should be sufficient to sync up the values. Changes to the cloud are buffered, so the system is optimized for what you're describing and you shouldn't need to worry about this kind of switching at all, the system is already designed to handle it for you.

I'm not fully sure what do you mean by "local version of cloud variables" though? Cloud variables are perfectly ok to use even if it's just the user syncing their own stuff.
